detection_rules/rule.py

```diff
@@ -364,38 +366,18 @@ def validate_query(self, meta: RuleMeta) -> None:
 AnyRuleData = Union[QueryRuleData, EQLRuleData, MachineLearningRuleData, ThresholdQueryRuleData, ThreatMatchRuleData]
 
 
-@dataclass(frozen=True)
-class TOMLRuleContents(MarshmallowDataclassMixin):
-    """Rule object which maps directly to the TOML layout."""
-    metadata: RuleMeta
-    data: AnyRuleData = field(metadata=dict(data_key="rule"))
-
-    @classmethod
-    def all_rule_types(cls) -> set:
-        types = set()
-        for subclass in typing.get_args(AnyRuleData):
-            field = next(field for field in dataclasses.fields(subclass) if field.name == "type")
-            types.update(typing.get_args(field.type))
-
-        return types
-
-    @classmethod
-    def get_data_subclass(cls, rule_type: str) -> typing.Type[BaseRuleData]:
-        """Get the proper subclass depending on the rule type"""
-        for subclass in typing.get_args(AnyRuleData):
-            field = next(field for field in dataclasses.fields(subclass) if field.name == "type")
-            if (rule_type, ) == typing.get_args(field.type):
-                return subclass
-
-        raise ValueError(f"Unknown rule type {rule_type}")
+class BaseRuleContents(ABC):
```

Comment:
instead of forcing this into the hierarchy, you could make a typing.Protocol, which is just a way of having a duck typing restriction, without requiring a particular sublcass
https://docs.python.org/3/library/typing.html#typing.Protocol